statistics/feedback.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"bytes"
+	"sort"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+type feedback struct {
+	lower *types.Datum
+	upper *types.Datum
+	count int64
+}
+
+// QueryFeedback is used to represent the query feedback info. It contains the query's scan ranges and number of rows
+// in each range.
+type QueryFeedback struct {
+	tableID  int64
+	hist     *Histogram
+	feedback []feedback
+	expected int64
+	actual   int64
+	valid    bool
+	desc     bool
+}
+
+// NewQueryFeedback returns a new query feedback.
+func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool) *QueryFeedback {
+	if hist != nil && hist.Len() == 0 {
+		hist = nil
+	}
+	return &QueryFeedback{
+		tableID:  tableID,
+		valid:    true,
+		hist:     hist,
+		expected: expected,
+		desc:     desc,
+	}
+}
+
+// StoreRanges stores the ranges for update.
+func (q *QueryFeedback) StoreRanges(ranges []*ranger.NewRange) {
+	q.feedback = make([]feedback, 0, len(ranges))
+	for _, ran := range ranges {
+		q.feedback = append(q.feedback, feedback{&ran.LowVal[0], &ran.HighVal[0], 0})
+	}
+}
+
+// Invalidate is used to invalidate the query feedback.
+func (q *QueryFeedback) Invalidate() {
+	q.feedback = nil
+	q.hist = nil
+	q.valid = false
+	q.actual = -1
+}
+
+// Actual gets the actual row count.
+func (q *QueryFeedback) Actual() int64 {
+	if !q.valid {
+		return -1
+	}
+	return q.actual
+}
+
+// Hist gets the histogram.
+func (q *QueryFeedback) Hist() *Histogram {
+	return q.hist
+}
+
+// Counts returns the counts info for each range.
+func (q *QueryFeedback) Counts() []int64 {
+	counts := make([]int64, 0, len(q.feedback))
+	for _, fb := range q.feedback {
+		counts = append(counts, fb.count)
+	}
+	return counts
+}
+
+// Update updates the query feedback.
+func (q *QueryFeedback) Update(startKey kv.Key, counts []int64) {
```

Comment:
add some comments about what these input parameters represented for and how we use it in this function ?